Jk/issue4

jamesgyoung opened the pull request:
This pr changes the function `to_data_array_as_categorical_raster` to `to_dataset_as_raster`.

The function will now:
- rasterise a column and assign the minimum possible datatype.
- allow a variable length list of column names to be passed and rasterise each one.

e.g. 

```python
from sds_data_model import vector

my_dataset = (
    vector.VectorLayer.from_files(
        data_path = "Local_Authority_Districts_(December_2021)_GB_BGC/LAD_DEC_2021_GB_BGC.shp",
        data_kwargs = {},
        name = "lad",    
    )
   .to_tiles()
   .to_dataset_as_raster(
        columns = ["OBJECTID", "BNG_E", "SHAPE_Area"]
    )
)

```

Returns an `xarray.Dataset` consisting of three `xarray.DataArray`'s, using types `uint16`, `uint32` & `float64`.

s-flint-defra committed:
Add data type method

s-flint-defra committed:
.

s-flint-defra committed:
Final refinements for data type options and instantiate tiles in to_data_array_as_categorical_raster for use in subsequent functions.

s-flint-defra committed:
Add docstrings and update method/variable names.

jamesgyoung committed:
add suport for multiple attributes

jamesgyoung commented:
Change was made to constants.py (issue #14) that may need to be ignored/rejected https://github.com/Defra-Data-Science-Centre-of-Excellence/sds-data-model/blob/1e8d0eeaeb56c119210f77e1a7b936bd97c2299e/src/sds_data_model/constants.py#L4

jpd-defra commented on src/sds_data_model/vector.py:
```diff
@@ -111,21 +113,32 @@ def to_mask(
         return mask
 
-    def to_categorical_raster(
+    def to_raster(
```
Could you add a docstring here please?

s-flint-defra replied:
Apologies I missed this one out as I was on the way out - I've added a description in line with the others (to be refined at a later date).

jpd-defra commented on src/sds_data_model/vector.py:
```diff
@@ -285,17 +326,17 @@ def from_files(
-            metadata = json.load(f"{data_path}-metadata.json")
-        else:
-            metadata = Metadata.from_file(metadata_path)
+        # if not metadata_path:
```
If commented out, is this a hangover from development and could be removed?

s-flint-defra replied:
Possibly - but James and I weren't able to run the packing with this line uncommented. I think this  might be something for @EFT-Defra to decide on.

EFT-Defra replied:
I was being over optimistic here in I thinking that we would always have metadata and that the default would be a CSV-W style JSON side-car file. I think I need to change the function to make metadata optional but allow the caller to either read metadata from a file or manually supply a Metadata object.

jpd-defra commented on src/sds_data_model/_vector.py:
```diff
@@ -99,18 +100,25 @@ def _get_mask(
     gpdf: GeoDataFrame,
-    categorical_column: str,
+    column: str,
 ) -> Generator[Tuple[BaseGeometry, Any], None, None]:
```
To be consistent with other private functions such as _get_col_dtype, could a docstring be added to this function too please?

s-flint-defra replied:
I can but I would be guessing as I believe this is @EFT-Defra's method, I did however change the variable name to reflect the fact it now handles all column types.

EFT-Defra replied:
I'll add a docstring.

jpd-defra commented on src/sds_data_model/constants.py:
```diff
@@ -1,7 +1,7 @@
 from itertools import product
 from typing import Tuple
 
-BNG = "OSGB36 / British National Grid"
```
Could add a single comment describing what these constants are for - I'm guessing they define the common raster layer created by the model.

s-flint-defra replied:
As per some of the other comments -Happy to have a stab but other than adding a constant I did not create this file.

EFT-Defra replied:
I think we need to handle "OSGB36 / British National Grid" or "OSGB 1936 / British National Grid" so I'll change this accordingly.

I'll also add some comments explaining what these constants are.

jpd-defra reviewed (changes requested):
Looking good! I was reviewing mainly from an "understanding" point of view, as this is all pretty new to me. Adding a few more docstrings and comments as I've suggested should help those new to the package understand its mechanics.  

It may be out of the scope of this PR, but some way of intergrating a purely generated example to demonstrate the new rasterization potential would be helpful. Is there a way to embed an example in the help/description of the function where a dummy dataset is generated from scratch, and put through the function to show the utility. Just an idea, so don't worry if it's not feasible.

EFT-Defra committed:
Change parameter type from `list` to `List` in `to_dataset_as_raster` for compatibility with python 3.8

jpd-defra committed:
Modify `List` type in `to_dataset_as_raster`

Removed the `...` from the `List` type, as it is not valid. Left `str` in place.

EFT-Defra committed:
Add comments and docstrings

EFT-Defra committed:
Check CRS strings is in sequence of BNG strings

EFT-Defra committed:
Make metadata optional

jpd-defra reviewed (approved):
Thanks @JamesKeny0n, @s-flint-defra & @EFT-Defra for your input. Hopefully this helps users understand the different components of the data model a bit more. Looking forward to using the new rasterise function.
